File: 13. B-Trees and Red-Black Trees - Exercises/Hierarchy - Author's solution/Java-Solution/src/main/Hierarchy.java
package main;

import java.util.*;
import java.util.stream.Collectors;

public class Hierarchy<T> implements IHierarchy<T> {

    private Node<T> root;

    private LinkedHashMap<T,Node<T>> nodes;

    public Hierarchy(T element){
        Node<T> elementNode = new Node<T>(element);
        this.setRoot(elementNode);
        this.nodes = new LinkedHashMap<>();
        this.nodes.put(element,elementNode);
    }

    private Node<T> getRoot() {
        return root;
    }

    private void setRoot(Node<T> root) {
        this.root = root;
    }

    public void Add(T parent, T child){
        if(!this.nodes.containsKey(parent)){
            throw new IllegalArgumentException("Parent does not exist!");
        }

        if(this.nodes.containsKey(child)){
            throw new IllegalArgumentException("Child element cannot have 2 parents!");
        }

        Node<T> parentNode = this.nodes.get(parent);
        Node<T> childNode = new Node<T>(child,parentNode);
        parentNode.getChildren().add(childNode);
        this.nodes.put(child,childNode);
    }

    public int getCount() {
        return this.nodes.size();
    }

    public void Remove(T element){
        if(!this.nodes.containsKey(element)){
            throw new IllegalArgumentException("Element does not exist!");
        }
        Node<T> elementNode = this.nodes.get(element);
        if(elementNode.getParent() == null){
            throw new IllegalStateException("Cannot delete the Root!");
        }

        Node<T> parent = elementNode.getParent();
        List<Node<T>> children = elementNode.getChildren();
        for (Node<T> child : children) {
            child.setParent(parent);
        }

        parent.getChildren().addAll(children);

        this.nodes.remove(element);
        parent.getChildren().remove(elementNode);
    }

    public boolean Contains(T element){
        return this.nodes.containsKey(element);
    }

    public T GetParent(T element){
        if(!this.nodes.containsKey(element)){
            throw new IllegalArgumentException("Element does not exist!");
        }

        Node<T> elementNode = this.nodes.get(element);
        if(elementNode.getParent() == null){
            return null;
        }

        return elementNode.getParent().getValue();
    }

    public Iterable<T> GetChildren(T element){
        if(!this.nodes.containsKey(element)){
            throw new IllegalArgumentException("Element does not exist!");
        }

        Node<T> elementNode = this.nodes.get(element);
        return elementNode.getChildren().stream().map(x->x.getValue()).collect(Collectors.toList());
    }

    public Iterable<T> GetCommonElements(IHierarchy<T> other){
        List<T> result = new ArrayList<T>();
        for (T key : this.nodes.keySet()) {
            if(other.Contains(key)){
                result.add(key);
            }
        }

        return result;
    }

    @Override
    public Iterator<T> iterator() {
        TreeIterator<T>it = new TreeIterator<>(getRoot());
        return it;
    }
}

File: 13. B-Trees and Red-Black Trees - Exercises/Hierarchy - Author's solution/C#-Solution/Hierarchy.Core/Hierarchy.cs
namespace Hierarchy.Core
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Collections;
    using System.Diagnostics;
    
    public class Hierarchy<T> : IHierarchy<T>
    {
        private Node<T> root;
        private Dictionary<T, Node<T>> nodes;

        public Hierarchy(T root)
        {
            this.root = new Node<T>(root);
            this.nodes = new Dictionary<T, Node<T>>();
            this.nodes.Add(root, this.root);
        }

        public int Count
        {
            get
            {
                return this.nodes.Count;
            }
        }

        public void Add(T element, T child)
        {
            Node<T> parentNode;
            if (!this.nodes.TryGetValue(element, out parentNode))
            {
                throw new ArgumentException("Parent element does not exist in hierarchy");
            }
            
            if (this.nodes.ContainsKey(child))
            {
                throw new ArgumentException("Child cannot have 2 parents");
            }
            
            var childNode = new Node<T>(child, parentNode);
            this.nodes[child] = childNode;
            parentNode.Children.Add(childNode);
        }

        public void Remove(T element)
        {
            Node<T> node;
            if (! this.nodes.TryGetValue(element, out node))
            {
                throw new ArgumentException("Element does not exist in hierarchy");
            }

            if (node.Parent == null)
            {
                throw new InvalidOperationException("Cannot delete root element");
            }

            node.Parent.Children.AddRange(node.Children);
            foreach (var child in node.Children)
            {
                child.Parent = node.Parent;
            }

            this.nodes.Remove(element);
            node.Parent.Children.Remove(node);
        }

        public IEnumerable<T> GetChildren(T item)
        {
            Node<T> node;
            if (!this.nodes.TryGetValue(item, out node))
            {
                throw new ArgumentException("Element does not exist in the hierarchy");
            }

            return node.Children.Select(n => n.Value);
        }

        public T GetParent(T item)
        {
            Node<T> node;
            if (!this.nodes.TryGetValue(item, out node))
            {
                throw new ArgumentException("Element does not exist in the hierarchy");
            }

            return node.Parent != null ? 
                node.Parent.Value : default(T);
        }

        public bool Contains(T value)
        {
            return this.nodes.ContainsKey(value);
        }

        public IEnumerable<T> GetCommonElements(Hierarchy<T> other)
        {
            var commonElements = new List<T>();
            foreach (T key in this.nodes.Keys)
            {
                if (other.nodes.ContainsKey(key))
                {
                    commonElements.Add(key);
                }
            }

            return commonElements;
        } 

        public IEnumerator<T> GetEnumerator()
        {
            if (this.Count == 0)
                yield break;

            var queue = new Queue<Node<T>>();
            queue.Enqueue(this.root);
            while (queue.Count > 0)
            {
                var current = queue.Dequeue();
                yield return current.Value;

                foreach (var subordinate in current.Children)
                {
                    queue.Enqueue(subordinate);
                }
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
}